enhancements/microshift/update-and-rollback.md

```diff
@@ -0,0 +1,180 @@
+
+## Proposal
+
+### Workflow Description
+
+The Operating System of an Edge Device running MicroShift will be created using
+Image Builder. This tool will allow `device administrators` to customize an edge
+optimized version of RHEL by adding RPMs and configuration of choice as described
+[here](https://github.com/openshift/enhancements/blob/master/enhancements/microshift/kubernetes-for-device-edge.md).
+
+When a new version of MicroShift needs to be rolled out, a new image shall be created
+and exposed to the edge devices. When working with thousands of Edge Devices,
+automatic detection and enforcement of upgrades is required.  Hence, Edge Devices will
+automatically detect the new exposed OSTree and perform a transactional upgrade.
+
+Before pivoting to the new Operating System Tree, a data backup needs to be performed
+in order to save the state of the running Edge Device. Due to the resource constrained
+environments where these Edge Devices run on, only the N-1 state will be supported. That 
+means that only one backup will be stored and when the upgrade is performed, the Edge
+Device can only rollback to the previous state N-1.
+
+The backup should consider all data required to save the state of MicroShift such as
+certificates, configuration files, volumes
+
+At every boot, `Greenboot` healthchecks will be executed as described in [this enhancement.](https://github.com/openshift/enhancements/blob/master/enhancements/microshift/microshift-greenboot.md)
+
+If upgrading MicroShift to a new version results in a broken state, `Greenboot` scripts
+will restore the previously saved backup, and perform an atomic rollback to the N-1 state.
+
+### API Extensions
+
+None
+
+### Implementation Details/Notes/Constraints [Optional]
+
+#### rpm-ostree Configuration
+
+`rpm-ostree` provides a configuration option to check and stage new updates.
+The file `/etc/rpm-ostreed.conf` provides `AutomaticUpdatePolicy` which defaults to `none`.
+
+This option needs to be set to `stage` which downloads and unpacks the update performing
+any package layering. Only a small amount of work is left to be performed at shutdown time
+via the `ostree-finalize-staged.service` systemd unit.
+
+Furthermore, in order to apply this policy, the `rpm-ostreed-automatic.service` systemd
+service must be enabled.
+
+#### MicroShift's Upgrade Systemd Service
+
+A new systemd service file will be packaged into the MicroShift's main RPM. This systemd
+service will monitor the `/run/ostree/staged-deployment` file that stores the staged update deployment info by using the `"ConditionPathExists"` option. This condition will be the
+trigger to run a custom script that saves the MicroShift state into a data backup file.
+
+#### MicroShift Backup
+
+Leveraging the ability to run any custom script before the new OSTree deployment becomes
+active, allow us to perform the following tasks in order:
+* Stop MicroShift systemd service (which does not cause interruption for applications)
```

Comment:
shall it estimate the memory/disk usage for executing backup cmds?

Reply:
The script/tool/code in charge of doing the backup should estimate if the backup will not get the system out of disk space, that's a good point. I'll add that.

In terms of memory, stopping MicroShift I guess it's good enough to avoid OOM issues.